### PR TITLE
Cluster Verification Framework

### PR DESCRIPTION
**Pull request by jayunit100**
I've spent the last few days looking at the general patterns of verification we have that we tend to reuse in the e2es.  Basically, we need 
- label filters
- forEach and WaitFor (where forEach doesn't necessarily waitFor anything).
- timeouts
- multiple phases (reusable definition of state)
- an extensible way to define cluster state that can evolve over time in a data object rather than as a set of parameters that have magic semantics

This PR 
- implements the abstract above functionality declaratively, and w/o hidden semantics.
- addresses the sprawling duplicate methods in #23540, so that we can phase out the wrapper methods and replace them with well defined, extensible semantics for cluster state.
- fixes the recently discovered #23730 issue (where kubectl.go is relying on examples.go, which is obviously wacky) by using the new framework to implement forEachPod in just a couple of lines and migrating the wrapper function into framework.go.

There is some cleanup to do here, but this is seemingly working for a couple of use cases that are important (spark,cassandra,...,kubectl) tests. - i played with a few different ideas and this wound up seeming to be the most natural implementation from a usability standpoint... 

in any case, just thought id push this up as a first iteration, open to feedback.

@kubernetes/sig-testing @timothysc

### CODE REVIEW

File: test/e2e/kubectl.go [hunk 1]
```diff
@@ -112,6 +112,14 @@ var (
 )
 
 var _ = KubeDescribe("Kubectl client", func() {
+	// localized helper function, not meant to be reused outside of this test.
+	forEachPod := func(framework *Framework, selectorKey string, selectorValue string, fn func(api.Pod)) {
```
Comment (jayunit100) [inline test/e2e/kubectl.go]: Rather than making ambiguous wrappers hidden in random tests, if a test needs syntactic sugar for a reused ClusterVerification, they can just declare it locally in 5 lines of code at the top of their test.

**Comment by timothysc** [comment]
It's on my radar to review shortly.

File: test/e2e/examples.go [hunk 2]
```diff
@@ -84,13 +90,13 @@ var _ = KubeDescribe("[Feature:Example]", func() {
 
 			By("checking up the services")
 			checkAllLogs := func() {
-				forEachPod(c, ns, "name", "redis", func(pod api.Pod) {
+				forEachPod("name", "redis", func(pod api.Pod) {
```
Comment (timothysc) [inline test/e2e/examples.go]: why was c & ns trim'd out of the signature.  Seems like they are still leveraged in the function are now relying on scope.

Comment (jayunit100) [inline test/e2e/examples.go]: entirely on scope.  yes.  This is a local wrapper , no need for `forEachPod` global wrapper, this fixes #23730 .

File: test/e2e/framework.go [hunk 3]
```diff
@@ -419,3 +420,136 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
 func KubeDescribe(text string, body func()) bool {
 	return Describe("[k8s.io] "+text, body)
 }
+
+// PodStateVerification represents a verification of pod state.
+// Any time you have a set of pods that you want to operate against or query,
+// this struct can be used to declaratively identify those pods.
+type PodStateVerification struct {
+	// Optional: only pods that have k=v labels will pass this filter.
+	selectors map[string]string
```
Comment (timothysc) [inline test/e2e/framework.go]: Isn't this whole thing essentially a SELECT function, with some of the WHERE clauses in place?    I'm just wondering if we can reduce further, or breakdown selector + phase + name as a select-filter sub-struct.

File: test/e2e/framework.go [hunk 4]
```diff
@@ -419,3 +420,136 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	}
+	// Next: Pod must match at least one of the states that the user specified
+	for _, pod := range pods {
+		// First Filter: Required
+		podStatePassesFilter := false
+		for _, phase := range p.validPhases {
+			if phase == pod.Status.Phase {
+				podStatePassesFilter = true
+			}
+		}
```
Comment (timothysc) [inline test/e2e/framework.go]: can't you just continue is podStatePassesFilter ==false here?

File: test/e2e/framework.go (hunk 3 above)
Comment (jayunit100) [inline test/e2e/framework.go]: the reason i didnt break it down further, was because this way you can do `PodStateVerification{map[string]string{selectorKey: selectorValue}..., []api.PodPhase{api.PodRunning}` super concise state definitions for the most common use cases.  maybe modularize further in a future iteration?

File: test/e2e/framework.go [hunk 5]
```diff
@@ -419,3 +420,136 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+		// Second filter
+		podFunctionVerified := p.verify == nil
+		if !podFunctionVerified {
+			podFunctionVerified, err = p.verify(pod)
+			if err != nil {
+				return podFiltered, err
+			}
+		}
```
Comment (timothysc) [inline test/e2e/framework.go]: same.

File: test/e2e/framework.go (hunk 4 above)
Comment (jayunit100) [inline test/e2e/framework.go]: we have to check all valid phases so can't have to have the inner for loop ?unless im misunderstanding..

File: test/e2e/framework.go [hunk 6]
```diff
@@ -419,3 +420,136 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	// this function allows specifying an error as well.
+	// The error indicates that the polling of the pod spectrum should stop.
+	verify func(api.Pod) (bool, error)
+
+	// Optional: only pods with this name will pass the filter.
+	podName string
+}
+
+type ClusterVerification struct {
+	framework *Framework
```
Comment (timothysc) [inline test/e2e/framework.go]: Please describe why namespace is no longer needed given that framework implies namespace.

File: test/e2e/framework.go (hunk 3 above)
Comment (jayunit100) [inline test/e2e/framework.go]: per tims update above.  

- i think there is value in this, 
- but after hacking around with it for a little while, though, i think it should be done in a later iteration.
- why? as-is, the separation between the zeroOrderPredicates and verify felt a little awkward after i implemented it into the API.  It added a little indirection.  

i think we should have a full blown generic state definition for kube, and this will be the seed for it , but i don't think this PR is the place to design the final product....  thoughts @timothysc ?

File: test/e2e/framework.go (hunk 6 above)
Comment (jayunit100) [inline test/e2e/framework.go]: this comment is insightful b/c actually its better to decouple the ClusterVerification from framework entirely, so that Namespace is explicit, and no higher reliance on e2e package is enforced.
I've done that and updated.

File: test/e2e/framework.go [hunk 7]
```diff
@@ -419,3 +420,166 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+
+func (f *Framework) NewClusterVerification(filter PodStateVerification) *ClusterVerification {
+	return &ClusterVerification{
+		f.Client,
+		f.Namespace.Name,
+		filter,
+	}
+}
+
+func passesPodNameFilter(pod api.Pod, name string) bool {
```
Comment (jayunit100) [inline test/e2e/framework.go]: @timothysc thes functions will hold us over until we make a more well-informed data model in #23913

Comment (timothysc) [inline test/e2e/framework.go]: I don't know how I feel about the name "passes", but I like the separation.  b/c what you are doing really is checking to see if objA ~= objB.  It's almost like string matching comparison.  

(match/compare/someothername)PodName(a, b)

File: test/e2e/framework.go [hunk 8]
```diff
@@ -419,3 +420,166 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	}
+
+	unfilteredPods := pl.Items
+	filteredPods := []api.Pod{}
+	if len(p.validPhases) == 0 {
+		panic("Need to specify a valid pod phase as part of pod states.")
+	}
+	// Next: Pod must match at least one of the states that the user specified
+	for _, pod := range unfilteredPods {
+		podStatePassesFilter := passesStateFilter(pod, p.validPhases)
```
Comment (timothysc) [inline test/e2e/framework.go]: much cleaner!

File: test/e2e/framework.go [hunk 9]
```diff
@@ -419,3 +420,166 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+
+		podFunctionVerified, err := passesVerifyFilter(pod, p.verify)
+		if err != nil {
+			return filteredPods, err
+		}
+
+		podNameVerified := passesPodNameFilter(pod, p.podName)
+		if err != nil {
```
Comment (timothysc) [inline test/e2e/framework.go]: There is no error return here, why is this needed?

File: test/e2e/framework.go (hunk 8 above)
Comment (timothysc) [inline test/e2e/framework.go]: I still think you could add the continues on the failed filters.

File: test/e2e/framework.go [hunk 10]
```diff
@@ -419,3 +420,166 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	}
+	return filteredPods, nil
+}
+
+// WaitFor waits for atLeast some number of pods to be verified, according to the PodStateVerification
+// definition.
+func (cl *ClusterVerification) WaitFor(atLeast int, timeout time.Duration) ([]api.Pod, error) {
+	var pods []api.Pod
+	var err error
+	for t := time.Now(); time.Since(t) < timeout; time.Sleep(1 * time.Second) {
```
Comment (timothysc) [inline test/e2e/framework.go]: This looks just like wait.Poll.. why not use that?

File: test/e2e/framework.go (hunk 8 above)
Comment (jayunit100) [inline test/e2e/framework.go]: we could bug logging gets hard to debug if we continue

**Comment by dchen1107** [comment]
cc/ me

File: test/e2e/framework.go (hunk 9 above)
Comment (jayunit100) [inline test/e2e/framework.go]: Yes, i removed err entirely from podNameVerified - this is now redundant.

**Comment by jayunit100** [comment]
@timothysc updated `Poll` functionality and the superfluous err check you caught (thanks!)

File: test/e2e/framework.go (hunk 7 above)
Comment (jayunit100) [inline test/e2e/framework.go]: "passes" only for uniformity w/ the other filter funcs.

File: test/e2e/framework.go [hunk 11]
```diff
@@ -419,3 +421,171 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+		if err != nil {
+			return false, err
+		} else {
+			return verified, nil
+		}
+
+	}
+}
+
+func passesStateFilter(pod api.Pod, validPhases []api.PodPhase) bool {
```
Comment (spiffxp) [inline test/e2e/framework.go]: maybe rename to `passesPhaseFilter`?  parity with `validPhases`

**Comment by timothysc** [comment]
@jayunit100 k I'll look-over one last time once the tests are passing.  We could needle this, but it's strictly better then what was there before.

File: test/e2e/framework.go [hunk 12]
```diff
@@ -419,3 +421,172 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	}
+
+	unfilteredPods := pl.Items
+	filteredPods := []api.Pod{}
+	if len(p.validPhases) == 0 {
+		panic("Need to specify a valid pod phase as part of pod states.")
+	}
+	// Next: Pod must match at least one of the states that the user specified
+	for _, pod := range unfilteredPods {
+		podStatePassesFilter := passesPhasesFilter(pod, p.validPhases)
```
Comment (timothysc) [inline test/e2e/framework.go]: you're still missing the `continue` on failed filters.

File: test/e2e/framework.go [hunk 13]
```diff
@@ -419,3 +421,172 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+			return filteredPods, err
+		}
+
+		podNameVerified := passesPodNameFilter(pod, p.podName)
+
+		// Compose them all...
+		if podStatePassesFilter && podNameVerified && podFunctionVerified {
+			filteredPods = append(filteredPods, pod)
+		} else {
+			Logf("pod %v state, name, verified %v, %v, %v ",
```
Comment (timothysc) [inline test/e2e/framework.go]: This is probably not needed.

File: test/e2e/framework.go [hunk 14]
```diff
@@ -419,3 +421,172 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	pl, err := filterLabels(p.selectors, c, ns) // Build an api.PodList to operate against.
+	Logf("Selector matched %v pods for %v=%v", len(pl.Items), p.selectors)
+	if len(pl.Items) == 0 || err != nil {
+		return pl.Items, err
+	}
+
+	unfilteredPods := pl.Items
+	filteredPods := []api.Pod{}
+	if len(p.validPhases) == 0 {
+		panic("Need to specify a valid pod phase as part of pod states.")
```
Comment (timothysc) [inline test/e2e/framework.go]: panic?  why not error?

File: test/e2e/framework.go [hunk 15]
```diff
@@ -419,3 +421,172 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+	return pl, err
+}
+
+// filter pods passing a filter.  It can be used to compose
+// the more useful abstractions like ForEach, WaitFor, and so on, which
+// can be used directly by tests.
+func (p *PodStateVerification) filter(c *client.Client, namespace *api.Namespace) ([]api.Pod, error) {
+	ns := namespace.Name
+	pl, err := filterLabels(p.selectors, c, ns) // Build an api.PodList to operate against.
+	Logf("Selector matched %v pods for %v=%v", len(pl.Items), p.selectors)
```
Comment (timothysc) [inline test/e2e/framework.go]: This seems more like a debug level log, I would expect we don't care to see it otherwise.

File: test/e2e/framework.go [hunk 16]
```diff
@@ -419,3 +421,172 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
...
+func (cl *ClusterVerification) WaitFor(atLeast int, timeout time.Duration) ([]api.Pod, error) {
+	var pods []api.Pod
+	var returnedErr error
+
+	err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
+		pods, returnedErr = cl.podState.filter(cl.client, cl.namespace)
+
+		// Failure
+		if returnedErr != nil {
+			Logf("Cutting polling short: We got an error from the pod filtering layer.")
```
Comment (timothysc) [inline test/e2e/framework.go]: I would set most of the logging in this to be debug only.

File: test/e2e/framework.go (hunk 13 above)
Comment (jayunit100) [inline test/e2e/framework.go]: So this is related to above. I actually feel like we shouldn't continue because it's useful to see this, given the declarative nature of things. If we continue this is useless....  But if we don't continue then every polling iteration prints status. The current other tests do something similar, printing status when polling / so that progress can be monitored. That's even more important since in he declarative framework users will have even further separation from the test feedback.

I can remove this but just clarifying that the reason we have this is because i want to report the entire pod matching state every time I poll, for things like spark this is useful where there are 2 stages to matching (state, and then log grep matching).

File: test/e2e/framework.go (hunk 15 above)
Comment (jayunit100) [inline test/e2e/framework.go]: I think it's important because without this in a battery of failed tests it's not clear if labels weren't found or if pod simply failed

File: test/e2e/framework.go (hunk 14 above)
Comment (jayunit100) [inline test/e2e/framework.go]: Yeah let me see what the convention is. Panic is a little strong ......

File: test/e2e/framework.go (hunk 13 above)
Comment (timothysc) [inline test/e2e/framework.go]: per-pod logging will be too much on large tests.  imagine the amount of logging that would be put out on a test cluster of size 1k+ w/ 250 pods/node and checking state.

Comment (timothysc) [inline test/e2e/framework.go]: Rolled up logging at a macro level is fine, but this is per-pod.

File: test/e2e/framework.go (hunk 15 above)
Comment (timothysc) [inline test/e2e/framework.go]: This is high enough up the stack that I'm ok with this one.

File: test/e2e/framework.go (hunk 13 above)
Comment (jayunit100) [inline test/e2e/framework.go]: yeah, at 1000 node scale this will get ludicrious i hadnt thought of that.

File: test/e2e/framework.go (hunk 16 above)
Comment (jayunit100) [inline test/e2e/framework.go]: eliminated most logs (except this one), should be pretty clean now.

**Comment by jayunit100** [comment]
Ok.  Pushed some minimization of logs except rollups.... i think close to convergence.
kept `panic` but moved it to the top of the filter function, since those errors are very deep and should force execution to die immediately (unless we don't want to use panic at all?).

**Comment by jayunit100** [comment]
squashed and removed excess logs.

**Comment by timothysc** [comment]
LGTM.

**Comment by jayunit100** [comment]
@k8s-bot test this please

**Comment by jayunit100** [comment]
@k8s-bot test this please

**Comment by jayunit100** [comment]
There is some definite flakiness going on thats unrelated to this patch.  I improved the logging a little and pushed a couple minor cleanups, needs another final stamp @timothysc
